### ISSUE TITLE
einsum implementation

### ISSUE THREAD

**Issue by aremirata** [pydata/sparse#31]
Hi guys, do you have plans to add einsum to this library? I guess einsum from numpy  is obtained from c. There are einsum operation I cannot directly convert to using tensordot. by the way, this library is really awesome and thanks for creating this.

**Comment by mrocklin** [comment]
I have no plans personally to add einsum, but it seems well within scope if
you're interested in implementing it.

**Comment by bigerl** [comment]
I realized that what I really need for my project is a highly parallelized einsum. I'm implementing it now as stand-alone and when I'm done I'll check out if it is a reasonable effort to integrate it with this library again. When it's done it will be open sourced anyways. I'll report back when I've got something new.

**Comment by hameerabbasi** [comment]
@aremirata @bigerl I've found in my experience that a lot of the stuff that's done using `einsum` is doable using broadcasting, reductions and `tensordot`. Maybe you can try those now that they're added?

**Comment by nils-werner** [comment]
I've written a very simple proof of concept: https://github.com/nils-werner/sparse/tree/einsum

It only works for two operands and doesn't support any of the fancy einsum tricks like ellipses etc.

Internally it uses tensordot and the fact that we can cheaply diagonalize axes on sparse arrays (see the docstrings for more explanation).

**Comment by nils-werner** [comment]
I've written a few small issues and added [some tests](https://github.com/nils-werner/sparse/blob/einsum/sparse/tests/test_core.py#L821-L842).

**Comment by hameerabbasi** [comment]
@nils-werner It'd be great to have this in, even with partial support. More features can always be added. We can document the missing features.

One minor thing: I'd really prefer to have this support any number of arrays rather than two.

**Comment by stsievert** [comment]
I'm using [opt_einsum] with their sparse backend, but opt_einsum still wants an einsum operation in sparse. It works for particular einsum strings as per [their sparse tests][1], but still relies an einsum operation for higher dimensions/more inputs.

The einsum strings that are throwing errors are `'bz,az->abz'`, `'cz,bz,az->abcz'` and `'dz,cz,bz,az->abcdz'` (and so on; always N two indices to one length N+1 index).


[1]:https://github.com/dgasmith/opt_einsum/blob/960cf7c45993077be6d4ad04a1564b6042f0e092/opt_einsum/tests/test_backends.py#L131-L158
[opt_einsum]:https://optimized-einsum.readthedocs.io/en/latest/

**Comment by hameerabbasi** [comment]
It'd be nice to have some concrete applications before we dive into this. I have a rough sketch of what a generalized implementation will look like, but it's way too complex for me to put into code at this point.

**Comment by stsievert** [comment]
Here's a failing test: https://gist.github.com/stsievert/202a07432ce82579a449ad113b33ba4a. It's takes the einsum string as input, but only fails for certain inputs. Its a modification of [the opt-einsum sparse tests](https://github.com/dgasmith/opt_einsum/blob/960cf7c45993077be6d4ad04a1564b6042f0e092/opt_einsum/tests/test_backends.py#L131-L158).

The application is with https://github.com/tensorly/tensorly/pull/64.

**Comment by hameerabbasi** [comment]
Okay, I’ll dive into it when I have time. This isn’t possible near-term though. You can look at the CSR matmul in the meantime and see if you can figure it out from that.

**Comment by zoome0215** [comment]
@jcmgray  I am doing scientific computation with einsum contraction of type ijsn,msk,tik,tjn ->tim with dimensions t=10000,i=1600,j=1600,s=n=m=k=25. the number of elements in msk is huge, but when I use the sparse matrix there are only 19 or so components.

It will be very useful if your `batched_contract` code can be expanded to multiple tensor inputs and merged to `sparse`.

It will be a great if sparse einsum is implemented; or any insights on this issue is given. Any news on development of einsum?

**Comment by nils-werner** [comment]
I have recently stumbled across [einops](https://github.com/arogozhnikov/einops) which [already supports a wide range of frameworks](https://github.com/arogozhnikov/einops#supported-frameworks). Maybe implementing an interface is a thing to consider?

**Comment by JackWetherell** [comment]
This is due to the fact that any string that would result in an output sparse array higher than 2D will fail. This is because einsum can produce arrays with a dimension larger than 2, and sparse does not support higher dimension arrays.

**Comment by hameerabbasi** [comment]
> I have recently stumbled across [einops](https://github.com/arogozhnikov/einops) which [already supports a wide range of frameworks](https://github.com/arogozhnikov/einops#supported-frameworks). Maybe implementing an interface is a thing to consider?

None of those have sparse support, sadly.

**Comment by smartass101** [comment]
I'd like to point out that without an `einsum` implementation one cannot directly leverage the full potential of this library when combined with `xarray`, even though `xarray` supports `sparse` containers. The reason is that [`xarray.dot`](https://xarray.pydata.org/en/stable/generated/xarray.dot.html) internally uses an `einsum` dispatcher. In my opinion `sparse` and `xarray` are a match made in heaven, but the lack of support for `xr.dot` currently hampers one of the main uses of this wonderful combination.

I tried to work around this with `xr.apply_ufunc` with something like this
```python
xr.apply_ufunc(sparse.tensordot, a, b, 
                      input_core_dims=[['dim1', 'dim_0'], ['dim_0']],
                      output_core_dims=[['dim_1']],
                      dask='parallelized',
                      output_dtypes=[b.dtype],
                      kwargs=dict(axes=(-1,-1)),
                     )
```
because `sparse.dot` uses the hardcoded `tensordot(a, b, axes=(-1, -2))`. But apparently `tensordot` silently squeezes dimensions, so this does not work either.

**Comment by hameerabbasi** [comment]
We are working on a re-write of PyData/Sparse in C++, that will, among other things, allow us to do blazing-fast einsums for sparse and dense arrays.

Only problem is, I no longer can do this at my day job currently, and it's highly theoretical, so progress is slow at the moment. Please follow along at https://github.com/hameerabbasi/xsparse.

**Comment by rabernat** [comment]
Big :+1: to implementation of einsum in sparse. As pointed out by @smartass101, this would really help with a lot of use  cases! I'm excited to see xsparse moving forward. 🚀

**Comment by finartist** [comment]
[opt_einsum](https://optimized-einsum.readthedocs.io/en/stable/) supports sparse

**Comment by jcmgray** [comment]
I've put an `einsum` implementation in #564, if anyone is interested testing it (either with regard to edge cases or performance).

**Comment by HadrienNU** [comment]
I would also be interested to this feature to be able to use sparse as a xarray backend. I have seen the pending pull request. Is there anything I can do to help in order to merge the pull request?

**Comment by hameerabbasi** [comment]
@HadrienNU I guess one thing would be to branch off the existing PR and help finish it off.

**Comment by jcmgray** [comment]
#564 has been merged if anyone wants to try `sparse.einsum` out now. There are so many use cases (especially once you add sparsity, sparse format, as additional parameters) that benchmarking is quite tricky. I'll just note for possible future optimization, that an alternative method to performing the core pairwise contraction operation is to leverage batch matrix multiplication - I do have a general implementation of this that works for `sparse` too.

**Comment by HadrienNU** [comment]
Thank you very much, I will try this out

**Comment by hameerabbasi** [comment]
Resolved by #564, if anything is missing we can open new issues for it.
